dll/win32/shell32/shlexec.cpp

```diff
@@ -519,20 +674,22 @@ static UINT_PTR SHELL_ExecuteW(const WCHAR *lpCmd, WCHAR *env, BOOL shWait,
             CloseHandle( info.hProcess );
         CloseHandle( info.hThread );
     }
-    else if ((retval = GetLastError()) >= 32)
+    else if ((retval = gle = GetLastError()) >= 32)
     {
         WARN("CreateProcess returned error %ld\n", retval);
         retval = ERROR_BAD_FORMAT;
     }
 
+done:
     TRACE("returning %lu\n", retval);
-
     psei_out->hInstApp = (HINSTANCE)retval;
 
     if (gcdret)
```

Comment:
I can't spot a single assignment currently where local variable gcdret would ever be set != 0 in this function.
When fixing this, make sure, that you will not introduce any code-branch that would leave the local variable gle uninitialized when it is accessed by RestoreLastError(gle); at line 691.

Reply:
gcdret probably stands for GetCurrentDirectoryReturn and it only gets set != 0 the one time `GetCurrentDirectoryW` gets called. Arguably the code is incorrect and you only need to restore the directory when both `GetCurrentDirectoryW` and `SetCurrentDirectoryW` succeed but that is beside the point

I can protect the call to `RestoreLastError` better if that would ease your mind.

Reply:
Saw it now at new line 628. For some reason my search didn't spot the gcdret-assignment yesterday. I do consider it being addressed now, since you did initialize gle now. I am not 100% sure whether there was a path through the program before, where it was used uninitialized, it was hard to follow. But I feel it to be more safe now. I do consider this topic being resolved now. Thx.